### PR TITLE
add gd32407v-start

### PR DESCRIPTION
**Pull request by Ouxiaolong**
## 拉取/合并请求描述：(PR description)

构建GD32系列BSP架构，新增BSP gd32407v-start。Build the GD32 series BSP architecture, and add the BSP gd32407v-start.

已经通过gd32407v-start开发板的测试验证。It has been tested and verified by the gd32407v-start development board.

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality
- [x] 本拉取/合并使用[formatting](https://github.com/mysterywolf/formatting)等源码格式化工具确保格式符合[RT-Thread代码规范](../documentation/coding_style_cn.md) This PR complies with [RT-Thread code specification](../documentation/coding_style_en.txt)

### CODE REVIEW

File: bsp/gd32/gd32407v-start/board/Kconfig [hunk 1]
```diff
@@ -0,0 +1,81 @@
+menu "Hardware Drivers Config"
+
+config SOC_GD32407V
+    bool 
+	select SOC_SERIES_GD32F4
+    select RT_USING_COMPONENTS_INIT
```
Comment (yangjie11) [inline bsp/gd32/gd32407v-start/board/Kconfig:6]: 注意格式对齐

File: bsp/gd32/gd32407v-start/board/SConscript [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+startup_path_prefix = SDK_LIB
+
+if rtconfig.CROSS_TOOL == 'gcc':
+    src += [startup_path_prefix + '/GD32F4xx_HAL/CMSIS/GD/GD32F4xx/Source/GCC/startup_gd32f4xx.S']
+elif rtconfig.CROSS_TOOL == 'keil':
+    src += [startup_path_prefix + '/GD32F4xx_HAL/CMSIS/GD/GD32F4xx/Source/ARM/startup_gd32f4xx.s']
+elif rtconfig.CROSS_TOOL == 'iar':
+    src += [startup_path_prefix + '/GD32F4xx_HAL/CMSIS/GD/GD32F4xx/Source/IAR/startup_gd32f4xx.s']
+
+CPPDEFINES = ['GD3232F407xx']
```
Comment (yangjie11) [inline bsp/gd32/gd32407v-start/board/SConscript:25]: 这个不对，你需要在厂商库文件中看到芯片定义

File: bsp/gd32/libraries/HAL_Drivers/drv_usart.c [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+}
+
+static const struct rt_uart_ops gd32_uart_ops =
+{
+    gd32_configure,
+    gd32_control,
+    gd32_putc,
+    gd32_getc,
+};
+
```
Comment (yangjie11) [inline bsp/gd32/libraries/HAL_Drivers/drv_usart.c:443]: static const struct rt_uart_ops _uart_ops =
{
    _uart_configure,
    _uart_control,
    _uart_putc,
    _uart_getc
};

File: bsp/gd32/libraries/HAL_Drivers/drv_gpio.c [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+const static struct rt_pin_ops gd32_pin_ops =
+{
+    gd32_pin_mode,
+    gd32_pin_write,
+    gd32_pin_read,
+    gd32_pin_attach_irq,
+    gd32_pin_detach_irq,
+    gd32_pin_irq_enable,
+    RT_NULL,
+};
```
Comment (yangjie11) [inline bsp/gd32/libraries/HAL_Drivers/drv_gpio.c:427]: 改为
```
    _pin_mode,
    _pin_write,
    _pin_read,
    _pin_attach_irq,
    _pin_detach_irq,
    _pin_irq_enable,
```

File: bsp/gd32/libraries/HAL_Drivers/drv_gpio.c [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+    index = get_pin(pin);
+    if (index == RT_NULL)
+    {
+        return RT_EINVAL;
+    }
+
+    hdr_index = bit2bitno(index->pin);
+    if (hdr_index < 0 || hdr_index >= ITEM_NUM(pin_irq_map))
+    {
+        return RT_EINVAL;
```
Comment (yangjie11) [inline bsp/gd32/libraries/HAL_Drivers/drv_gpio.c:322]: 除了RT_EOK之外，其他返回值添加负号

File: bsp/gd32/libraries/HAL_Drivers/Kconfig [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    depends on (SOC_SERIES_STM32L4 || SOC_SERIES_STM32F0 || SOC_SERIES_STM32F7 || SOC_SERIES_STM32H7 || SOC_SERIES_STM32MP1)
+    default n 
+
+config BSP_USING_RNG
+    bool "Enable RNG (Random Number Generator)"
+    select RT_USING_HWCRYPTO
+    select RT_HWCRYPTO_USING_RNG
+    depends on (SOC_SERIES_STM32L4 || SOC_SERIES_STM32F4 || SOC_SERIES_STM32F7 || \
+                SOC_SERIES_STM32H7 || SOC_SERIES_STM32MP1)
+    default n
```
Comment (yangjie11) [inline bsp/gd32/libraries/HAL_Drivers/Kconfig:42]: 这里不对，改一下

Comment (yangjie11) [reply bsp/gd32/libraries/HAL_Drivers/Kconfig:42]: image

File: bsp/gd32/gd32407v-start/board/SConscript (hunk 2 above)
Comment (yangjie11) [reply bsp/gd32/gd32407v-start/board/SConscript:23]: 这里没修改

File: bsp/gd32/gd32407v-start/board/SConscript [hunk 7]
```diff
@@ -19,8 +19,6 @@ if rtconfig.CROSS_TOOL == 'gcc':
     src += [startup_path_prefix + '/GD32F4xx_HAL/CMSIS/GD/GD32F4xx/Source/GCC/startup_gd32f4xx.S']
 elif rtconfig.CROSS_TOOL == 'keil':
     src += [startup_path_prefix + '/GD32F4xx_HAL/CMSIS/GD/GD32F4xx/Source/ARM/startup_gd32f4xx.s']
-elif rtconfig.CROSS_TOOL == 'iar':
-    src += [startup_path_prefix + '/GD32F4xx_HAL/CMSIS/GD/GD32F4xx/Source/IAR/startup_gd32f4xx.s']
```
Comment (yangjie11) [inline bsp/gd32/gd32407v-start/board/SConscript]: 这个不用删，是用来判断环境的

File: bsp/gd32/README.md [hunk 8]
```diff
@@ -0,0 +1,10 @@
+
+# GD32 BSP 说明
+
+GD32 系列 BSP 目前支持情况如下表所示：
+
+| **BSP 文件夹名称**       | **开发板名称**                 |
+|:------------------------- |:-------------------------- |
+| **F4 系列** |  |
+| [gd32407v-start](gd32407v-start) | 兆易创新 官方 GD32407V-START 开发板 |
```
Comment (yangjie11) [inline bsp/gd32/README.md:9]: 这个readme再完善一下

Comment (Ouxiaolong) [reply bsp/gd32/README.md:9]: 已经完善并再次提交了

**Pull request merged**